apps/src/gamelab/AnimationTab/PiskelEditor.jsx

```diff
@@ -24,47 +24,123 @@ const PiskelEditor = React.createClass({
+     *          events during load. */
+    this.isLoadingAnimation_ = false;
+
+    /** @private {AnimationKey} reference to animation that is currently loaded
+     *          in the editor. */
+    this.loadedAnimation_ = null;
+
+    this.piskel = new PiskelApi();
+    this.piskel.attachToPiskel(this.iframe);
+    this.piskel.onPiskelReady(this.onPiskelReady);
+    this.piskel.onStateSaved(this.onAnimationSaved);
+  },
+
+  componentWillUnmount() {
+    this.piskel.detachFromPiskel();
+    this.piskel = undefined;
   },
 
   componentWillReceiveProps(newProps) {
-    const {animations, selectedAnimation} = newProps;
     if (newProps.selectedAnimation !== this.props.selectedAnimation) {
-      this.postMessage({
-        type: MessageType.LOAD_ANIMATION,
-        animation: animations.find(animation => animation.key === selectedAnimation)
-      });
+      this.loadSelectedAnimation_(newProps);
     }
   },
 
-  componentShouldUpdate() {
+  loadSelectedAnimation_(props) {
+    const key = props.selectedAnimation;
+    if (!this.isPiskelReady_) {
+      return;
+    }
+
+    if (key === this.loadedAnimation_) {
+      // I wonder if this is ever valid - like we want to load some external edit?
+      return;
+    }
+
+    if (!key) {
+      // TODO: Put Piskel into a 'nothing-selected' state?
+      return;
+    }
+
+    if (this.isLoadingAnimation_) {
+      return;
+    }
+
+    const data = props.animationList.data[key];
+    if (!data) {
+      throw new Error('No data present for animation with key ' + key);
+    }
+
+    this.isLoadingAnimation_ = true;
+    this.piskel.loadSpritesheet(
+        data.dataURI,
+        data.frameSize.x,
+        data.frameSize.y,
+        data.frameRate,
+        () => {
+          this.loadedAnimation_ = key;
+          this.isLoadingAnimation_ = false;
+
+          // If the selected animation changed out from under us, load again.
+          if (this.props.selectedAnimation !== key) {
+            this.loadSelectedAnimation_(this.props);
+          }
+        });
+  },
+
+  // We are hosting an embedded application in an iframe; we should never try
+  // to re-render it.
+  shouldComponentUpdate() {
     return false;
   },
 
-  /**
-   * Send a message to Piskel.
-   * The message should be an object (much like a Redux action) that the Piskel
-   * API can handle.
-   * @param {object} message
-   */
-  postMessage(message) {
-    // Piskel should be hosted on the same origin (domain) as gamelab; we don't
-    // want to send messages to other domains.
-    // TODO (bbuchanan): Inject window origin to remove global dependency?
-    const targetOrigin = location.origin;
-    this.iframe.contentWindow.postMessage(message, targetOrigin);
+  onPiskelReady() {
+    this.isPiskelReady_ = true;
+    this.loadSelectedAnimation_(this.props);
+  },
+
+  onAnimationSaved(message) {
+    if (this.isLoadingAnimation_) {
+      return;
+    }
+    console.log('onPiskelSaveEvent', message); // TODO: Remove
```

Comment:
TODO: remove (not sure if needs to be removed immediately or not, but figure i'd mention in case you missed this)